working/0546-patterns/records-feature-specification.md

```diff
@@ -283,6 +283,41 @@ fields are) and collection literals.
 
 **TODO: Specify this more precisely.**
 
+### Constants
+
+A record expression is a compile-time constant expression
+if all its record field expressions are compile-time constant expressions. 
+
+_This is true whether the expression occurs in a constant context or not,
+which means that a record expression can be used directly as a parameter default value 
+if its record field expressions are constant expressions.
+Example: `f({(int, int) x = (1, 2)}) => ...`._
+
+A record expression is a potentially constant expression 
+if all its record field expressions are potentially constant or constant expressions.
+
+_This means that a record expression can be used in the initializer list
+of a constant non-redirecting generative constructor._
+
+Constant *object* instantiations create deeply immutable and canonicalied objects.
+Records are always unmodifiable, and if their field values are deeply immutable,
+like constants values, the records are also deeply immutable.
+It's meaningless to consider whether record constants are canonicalized,
+since records do not have a persistent identity.
+
+Because of that, there is no need for a `const (1, 2)` syntax to force a record 
+to be a constant, like there is for object creation expressions. 
+A record expression with field values that are constant-created values, 
+will be indistinguishable from a similar expression created in a constant 
+context, since identity cannot be used as a distinguishing trait.
+
+_(We could choose to promise that a compile-time constant `identical(c1, c2)`,
+where the expression occurs in a constant context and `c1` and `c2` are records, 
+will evaluate to `true` iff a runtime evaluation of `identical` 
+*can* return `true` for the same values. 
+That is, records would be canonicalized during compile-time constant evealuation,
+but may lose their identity at runtime. We will not make such a promise.)_
```

Comment:
I think such a promise would be too big, anyway. We would not _prevent_ unboxing from occurring on a canonicalized record, we would just allow canonicalization of certain record values to improve performance whenever possible.

```dart
const r0 = (1, 2);

void f((int, int) r) {
  ... // Uses `r.$0` and `r.$1`.
}
// A customized variant of `f` is generated by the compiler:
void f_unboxed(int r1, int r2) {
  ... // Copy of body of `f`, using `r1` and `r2`.
}

bool b = true;

void main() {
  var r1 = (Datetime.now().millisecondsSinceEpoch, 2);
  var r = b ? r0 : r1;
  f(r); // May compile to `f_unboxed(r.$0, r.$1)`.
}
```

Reply:
It'd only be a guarantee during compilation. When we perform an `identical(v1, v2)` *as a constant expression*, we have (representations of) the values available, and we can choose to perform the slower structural comparions each time, while still using the fast-path `identical` at runtime.

You won't be calling `f_unboxed` at compile-time.

Now, bigger question, should `record.$0` or `record.name` be constant expressions if `record` is?

Reply:
Part of https://github.com/dart-lang/language/issues/2219. ;-)